engine/common/requester/engine.go

```diff
@@ -426,8 +426,14 @@ func (e *Engine) onEntityResponse(originID flow.Identifier, res *messages.Entity
 		delete(needed, entityID)
 		delete(e.items, entityID)
 
-		// process the entity
-		go e.handle(originID, entity)
+		// validate that we got correct entity, exactly what we were expecting
+		if entityID == entity.ID() {
+			// process the entity
+			go e.handle(originID, entity)
+		} else {
+			// at this point we got an entry with invalid ID, so not the one that we were expecting
+			e.log.Error().Msgf("could not validate response body for requested entity %v", entityID)
```

Comment:
do we need to fail here with error?

Reply:
yeah, we need to return error, instead of logging, [just like what we did above](https://github.com/onflow/flow-go/pull/327/files#diff-61db1a3537aad3825bb345a268679ae434b54d503c67b619d5831598cc33a44fR402)

And we could handle the error case first:
```
actualEntityID := entity.ID()

if entityID != actualEntityID {
  return NewInvalidErrorf("expected id: %v, actual id: %v ...")
}

go e.handle(originID, entity)
```

Reply:
Sorry, why do we need to error here @zhangchiqing ? Here my thoughts on erroring: https://github.com/onflow/flow-go/pull/327#discussion_r568103932

Reply:
Ask you asking why do we need to `return` error?

1. The caller is logging the error, log here alone would break the connection between the call stack and this error. 
2. In terms of why return the error, instead of continue, my original thought was if we detected an invalid entity, the sender must be dishonest, so we should just stop and drop the rest. However, if we only consider non-BFT scenarios, both makes sense.

Reply:
Yes, I was asking why we need to error. Thanks Leo for the explanation. 


> * We process all entities until we hit the first error and then discard the remaining ones, even if they are valid. In my mind, it would be more consistent to either 
>     - **accept all valid entities and discard the invalid ones** (irrespective of order); or 
>     - **discard all entities if _at least one_ is invalid**. 

>  would break the connection between the call stack and this error.
* The caller is the networking layer, which only logs the error either. From the perspective of the networking layer, I would argue that feeding the requester engine with invalid inputs is part of the normal operations.